packages/flutter_tools/lib/src/compute_dev_dependencies.dart

```diff
@@ -2,126 +2,184 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'package:package_config/package_config.dart';
+import 'package:yaml/yaml.dart';
+
+import 'base/file_system.dart';
 import 'base/logger.dart';
-import 'convert.dart';
-import 'dart/pub.dart';
 import 'project.dart';
 
-/// Returns dependencies of [project] that are _only_ used as `dev_dependency`.
+/// Computes a representation of the transitive dependency graph rooted at
+/// [project].
+///
+/// Does a search rooted in `project`, following the `dependencies` and
+/// `dev_dependencies` of the pubspec to find the transitive dependencies of the
+/// app (including the project itself). Will follow the `dev_dependencies` of
+/// the [project] package.
+///
+/// Will load each of the dependencies' `pubspec.yaml` using [fileSystem]. Using
+/// [packageConfig] to locate the files.
+///
+/// Avoids loading the [project] manifest again.
 ///
-/// That is, computes and returns a subset of dependencies, where the original
-/// set is based on packages listed as [`dev_dependency`][dev_deps] in the
-/// `pubspec.yaml` file, and removing packages from that set that appear as
-/// dependencies (implicitly non-dev) in any non-dev package depended on.
-Future<Set<String>> computeExclusiveDevDependencies(
-  Pub pub, {
-  required Logger logger,
-  required FlutterProject project,
-}) async {
-  final Map<String, Object?> jsonResult = await pub.deps(project);
+/// If a pubspec cannot be read, or is malformed, that pubspec is skipped. If
+/// nothing has changed since a succesful `pub get` that should never happen.
+/// In case of `flutter run --no-pub` this will at most ignore new dependencies.
+Map<String, Dependency> computeTransitiveDependencies(
+  FlutterProject project,
+  PackageConfig packageConfig,
+  FileSystem fileSystem,
+  Logger logger,
+) {
+  final Map<String, Dependency> result = <String, Dependency>{};
+  result[project.manifest.appName] = Dependency(
+    project.manifest.toYaml(),
+    project.manifest.dependencies.toList(),
+    project.directory.uri,
+    isExclusiveDevDependency: true,
```

Comment:
I'm confused here, why is the project itself listed as a dev-dependency of itself?

If it's because *everything* starts as a dev dependency and then is changed, please leave a comment.

Reply:
Yes exactly! - we build the graph with everything marked as exclusive dev-dependencies, and then re-traverse the non-dev-dependencies part to remove that marker.

I'll add a comment!